typed/merge.go

```diff
@@ -140,81 +144,89 @@ func (w *mergingWalker) prepareDescent(pe fieldpath.PathElement, tr schema.TypeR
-	out := &value.List{}
+func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (errs ValidationErrors) {
+	rLen := 0
+	if rhs != nil {
+		rLen = rhs.Length()
+	}
+	lLen := 0
+	if lhs != nil {
+		lLen = lhs.Length()
+	}
+	out := make([]interface{}, 0, int(math.Max(float64(rLen), float64(lLen))))
 
 	// TODO: ordering is totally wrong.
 	// TODO: might as well make the map order work the same way.
 
 	// This is a cheap hack to at least make the output order stable.
-	rhsOrder := []fieldpath.PathElement{}
+	rhsOrder := make([]fieldpath.PathElement, 0, rLen)
 
 	// First, collect all RHS children.
-	var observedRHS fieldpath.PathElementValueMap
+	observedRHS := fieldpath.MakePathElementValueMap(rLen)
 	if rhs != nil {
-		observedRHS = fieldpath.MakePathElementValueMap(len(rhs.Items))
-		for i, child := range rhs.Items {
+		for i := 0; i < rhs.Length(); i++ {
+			child := rhs.At(i)
 			pe, err := listItemToPathElement(t, i, child)
 			if err != nil {
-				errs = append(errs, w.errorf("rhs: element %v: %v", i, err.Error())...)
+				errs = append(errs, errorf("rhs: element %v: %v", i, err.Error())...)
 				// If we can't construct the path element, we can't
 				// even report errors deeper in the schema, so bail on
 				// this element.
-				continue
+				return
 			}
 			if _, ok := observedRHS.Get(pe); ok {
-				errs = append(errs, w.errorf("rhs: duplicate entries for key %v", pe.String())...)
+				errs = append(errs, errorf("rhs: duplicate entries for key %v", pe.String())...)
 			}
 			observedRHS.Insert(pe, child)
 			rhsOrder = append(rhsOrder, pe)
 		}
 	}
 
 	// Then merge with LHS children.
-	var observedLHS fieldpath.PathElementSet
+	observedLHS := fieldpath.MakePathElementSet(lLen)
 	if lhs != nil {
-		observedLHS = fieldpath.MakePathElementSet(len(lhs.Items))
-		for i, child := range lhs.Items {
+		for i := 0; i < lhs.Length(); i++ {
+			child := lhs.At(i)
 			pe, err := listItemToPathElement(t, i, child)
 			if err != nil {
-				errs = append(errs, w.errorf("lhs: element %v: %v", i, err.Error())...)
+				errs = append(errs, errorf("lhs: element %v: %v", i, err.Error())...)
 				// If we can't construct the path element, we can't
 				// even report errors deeper in the schema, so bail on
 				// this element.
-				continue
+				return
 			}
 			if observedLHS.Has(pe) {
-				errs = append(errs, w.errorf("lhs: duplicate entries for key %v", pe.String())...)
-				continue
+				errs = append(errs, errorf("lhs: duplicate entries for key %v", pe.String())...)
+				return
 			}
 			observedLHS.Insert(pe)
 			w2 := w.prepareDescent(pe, t.ElementType)
-			w2.lhs = &child
+			w2.lhs = value.Value(child)
 			if rchild, ok := observedRHS.Get(pe); ok {
-				w2.rhs = &rchild
+				w2.rhs = rchild
 			}
 			if newErrs := w2.merge(); len(newErrs) > 0 {
-				errs = append(errs, newErrs...)
+				errs = append(errs, newErrs.WithPrefix(pe.String())...)
```

Comment:
Why is a WithPrefix merited here but not immediately above on line 217?

Reply:
because these errors have been created at a sub-level (within that pe) compared to the error above.